### PR TITLE
expression: check Overflow int128

### PR DESCRIPTION
**Pull request by fzhedu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/4512

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by fuzhe1989** [comment]
consider port https://github.com/ClickHouse/ClickHouse/blob/master/base/base/arithmeticOverflow.h to TiFlash?

**Review by yibin87 — COMMENTED** [review]
It'll be nice to explain why muloverflow affects comparasion result.

**Comment by fzhedu** [comment]
> consider port https://github.com/ClickHouse/ClickHouse/blob/master/base/base/arithmeticOverflow.h to TiFlash?

https://github.com/ClickHouse/ClickHouse/blob/master/base/base/arithmeticOverflow.h#L179
 /// Overflow check is not implemented for big integers.

but here needs to checkout overflow.

**Comment by fzhedu** [comment]
> It'll be nice to explain why muloverflow affects comparasion result.

a git positive value becomes negective, so the compare resultes are wrong.

**Review by yibin87 — APPROVED** [review]
LGTM

File: libs/libcommon/include/common/arithmeticOverflow.h [hunk 1]
```diff
@@ -113,9 +113,7 @@ inline bool mulOverflow(__int128 x, __int128 y, __int128 & res)
     if (!x || !y)
         return false;
 
-    unsigned __int128 a = (x > 0) ? x : -x;
-    unsigned __int128 b = (y > 0) ? y : -y;
-    return (a * b) / b != a;
+    return res/x != y; /// whether overflow int128
```
Comment (windtalker) [inline libs/libcommon/include/common/arithmeticOverflow.h:116]: How about add some unit test for this overflow check?

Comment (fzhedu) [reply libs/libcommon/include/common/arithmeticOverflow.h:116]: added

File: libs/libcommon/src/tests/gtest_arithmetic_overflow.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    Int256 res256;
+    /// 2^254
+    static constexpr Int256 int_254 = Int256((Int256(0x1) << 254));
+    /// 2^254 << 0 = 2^254
+    is_overflow = common::mulOverflow(int_254, Int256(1), res256);
+    ASSERT_EQ(is_overflow, false);
+
+    /// 2^254 << 1 = 2^255
+    is_overflow = common::mulOverflow(int_254, Int256(2), res256);
+    ASSERT_EQ(is_overflow, true);
```
Comment (fzhedu) [inline libs/libcommon/src/tests/gtest_arithmetic_overflow.cpp:48]: @LittleFall is this as expected?

Comment (LittleFall) [reply libs/libcommon/src/tests/gtest_arithmetic_overflow.cpp:48]: I think it is expected, the max limit of signed Int256  is (1<<255)-1

Comment (LittleFall) [reply libs/libcommon/src/tests/gtest_arithmetic_overflow.cpp:48]: According to the boost doc and implement, the `Int256` in boost includes **256bit data and 1bit sign**, so the max/min value is different from our expectation.

https://www.boost.org/doc/libs/1_60_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints/cpp_int.html

File: libs/libcommon/include/common/arithmeticOverflow.h [hunk 3]
```diff
@@ -129,6 +127,6 @@ inline bool mulOverflow(Int256 x, Int256 y, Int256 & res)
     {
         return true;
     }
-    return false;
+    return res >= Int256((Int256(0x1) << 255)); /// whether overflow int256
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply libs/libcommon/include/common/arithmeticOverflow.h:130]: > Why need add this?

please see https://github.com/pingcap/tiflash/pull/4762/files#r873432216

**Review by windtalker — APPROVED** [review]
LGTM
